### PR TITLE
{Util} Introduce a centralized function for subprocessing

### PR DESCRIPTION
**Pull request by jiasli**
**Related command**

**Description**
An alternative solution to https://github.com/Azure/azure-cli/pull/29503.

Judging by how `subprocess` is currently used:

https://github.com/Azure/azure-cli/blob/f1d068ece6789d8e77f3a1ee7c140b12982298a8/src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py#L105

https://github.com/Azure/azure-cli/blob/cff8d9bad3e39aa69ab9ca108693c66088d615bc/src/azure-cli/azure/cli/command_modules/resource/_bicep.py#L327-L331

https://github.com/Azure/azure-cli/blob/f4cf2963264145541cc10628af1743e2c93d669f/src/azure-cli/azure/cli/command_modules/mysql/_util.py#L388

https://github.com/Azure/azure-cli/blob/91aace5f851b5e70504efbb89079bb6170d09e74/src/azure-cli/azure/cli/command_modules/rdbms/_flexible_server_util.py#L331

we only need to expose limited arguments:

- `args` (positional)
- `stdin` (keyword)
- `stdout` (keyword)
- `check` (keyword)
- `env` (keyword)

It is not necessary to expose all `*args` and `**kwargs` from `subprocess.run`. 

This PR introduces a centralized function `cmd` for subprocessing with a more explicit and simplified signature.

Internally, to adhere to https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module, `cmd` uses the recommended function `subprocess.run`.

According to https://docs.python.org/3/library/subprocess.html#older-high-level-api, these old APIs are not exposed:

- `call`
- `check_call`
- `check_output`

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Util

File: src/azure-cli/azure/cli/command_modules/rdbms/_flexible_server_util.py [hunk 1]
```diff
@@ -318,18 +318,16 @@ def _resolve_api_version(client, provider_namespace, resource_type, parent_path)
...
+        process = cmd(commands, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
     if process.returncode:
         logger.warning(process.stderr.read().strip().decode('UTF-8'))
 
 
 def run_subprocess_get_output(command):
-    commands = command.split()
-    process = subprocess.Popen(commands, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    process.wait()
+    command = shlex.split(command)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/rdbms/_flexible_server_util.py:329]: Using a normal `split` is wrong, as it doesn't take quotes into consideration.

File: src/azure-cli/azure/cli/command_modules/mysql/_util.py [hunk 2]
```diff
@@ -374,19 +375,18 @@ def _resolve_api_version(client, provider_namespace, resource_type, parent_path)
 
 
 def run_subprocess(command, stdout_show=None):
+    command = shlex.split(command)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/mysql/_util.py:378]: The best practice is to construct `command` as a `list` from the very beginning.

File: src/azure-cli-core/azure/cli/core/util.py [hunk 3]
```diff
@@ -1356,3 +1356,23 @@ def should_encrypt_token_cache(cli_ctx):
     encrypt = cli_ctx.config.getboolean('core', 'encrypt_token_cache', fallback=fallback)
 
     return encrypt
+
+
+def run_cmd(args, *, capture_output=False, check=False, env=None):
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/util.py:1361]: `cmd` is a name widely used in the code base, so we use a more unique name.

File: src/azure-cli/azure/cli/command_modules/mysql/_util.py [hunk 4]
```diff
@@ -374,19 +374,18 @@ def _resolve_api_version(client, provider_namespace, resource_type, parent_path)
...
+        process = run_cmd(command, capture_output=True)
     if process.returncode:
         logger.warning(process.stderr.read().strip().decode('UTF-8'))
 
 
 def run_subprocess_get_output(command):
-    commands = command.split()
-    process = subprocess.Popen(commands, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    process.wait()
+    command = shlex.split(command)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/mysql/_util.py:387]: Using a normal `split` is wrong, as it doesn't take quotes into consideration.

**Comment by AllyW** [comment]
The listed keyword (stdin, stdout, check, env ) is not enough for all the usecases in cli, like

 `universal_newlines` and `timeout` in `acs` module here: https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L295-L298, 

and `text` in https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1772. 

If not all use cases need to be covered, we'll discuss which ones are supported by cli.

File: src/azure-cli/azure/cli/command_modules/mysql/_util.py (hunk 2 above)
Comment (AllyW) [reply src/azure-cli/azure/cli/command_modules/mysql/_util.py:377]: Does the POSIX mode need to be mentioned therefore developers may notice the potential difference when splitting their cmds?

Agreed on constructing the list of cmd at the beginning

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/mysql/_util.py:377]: The `command` string is constructed by `mysql` module itself:

https://github.com/Azure/azure-cli/blob/9119731bf5940981c3fe1f65ae495699cdb4b019/src/azure-cli/azure/cli/command_modules/mysql/_util.py#L416

The developer of `mysql` should be aware they should follow POSIX convention to build the `command` string.

**Comment by jiasli** [comment]
> `universal_newlines` and `timeout` in `acs` module here: https://github.com/Azure/azure-cli/blob/bb022828d41467ab8ac423c452bae74f5ef9a249/src/azure-cli/azure/cli/command_modules/acs/custom.py#L295-L297

`universal_newlines` is deprecated according to https://docs.python.org/3/library/subprocess.html#subprocess.run

> The `universal_newlines` argument is equivalent to `text` and is provided for backwards compatibility.

`timeout` can be added to `run_cmd`.

> and `text` in https://github.com/Azure/azure-cli/blob/bb022828d41467ab8ac423c452bae74f5ef9a249/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1772

I prefer not to complicate `run_cmd` too much. It is ok to let the caller `decode`s by itself:

**Comment by jiasli** [comment]
> `universal_newlines` and `timeout` in `acs` module here: https://github.com/Azure/azure-cli/blob/bb022828d41467ab8ac423c452bae74f5ef9a249/src/azure-cli/azure/cli/command_modules/acs/custom.py#L295-L297

`universal_newlines` is deprecated according to https://docs.python.org/3/library/subprocess.html#subprocess.run

> The `universal_newlines` argument is equivalent to `text` and is provided for backwards compatibility.

`timeout` can be added to `run_cmd`.

> and `text` in https://github.com/Azure/azure-cli/blob/bb022828d41467ab8ac423c452bae74f5ef9a249/src/azure-cli/azure/cli/command_modules/acs/custom.py#L1772

I prefer not to complicate `run_cmd` too much. It is ok to let the caller `decode` the output by itself, such as when calling `azcopy`:

https://github.com/Azure/azure-cli/blob/bd27913d0237806e2a4f2bf0d24a78351c58f63a/src/azure-cli/azure/cli/command_modules/storage/azcopy/util.py#L45-L46

**Comment by jiasli** [comment]
Superseded by #29745

**Pull request closed without merging**